lib/Scheduler/Scheduler.cpp

```diff
@@ -0,0 +1,116 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Scheduler/Scheduler.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "ttmlir/Dialect/TTNN/IR/TTNNOpsDialect.h.inc"
+#include <mlir/IR/BuiltinOps.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/Interfaces/DestinationStyleOpInterface.h>
+
+namespace mlir::tt::scheduler {
+
+bool isDpsOperand(mlir::Operation *op, const mlir::OpOperand *operand) {
+  // Check if the operation implements DestinationStyleOpInterface
+  if (auto dpsOp = dyn_cast<mlir::DestinationStyleOpInterface>(op)) {
+    if (dpsOp.getDpsInitOperand(0) == operand) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool isTTNNOp(mlir::Operation *op) {
+  return isa<ttnn::TTNNDialect>(op->getDialect());
+}
+
+bool isDpsOp(mlir::Operation *op) {
+  return isa<mlir::DestinationStyleOpInterface>(op);
+}
+
+// Constructor taking an MLIR Operation (or a module)
```

Comment:
It's not OR. Based on the assert you need to be clear in the comment that only ModuleOp is accepted.

Reply:
I think funcop is fine too. In case of module you need to allow multiple funcops.

Reply:
It easier for me to allow only func op from implementation perspective. Is that okay for you guys?